src/main/java/org/scoverage/plugin/SCoveragePreCompileMojo.java

```diff
@@ -444,22 +443,39 @@ private Artifact getScalaScoveragePluginArtifact( String resolvedScalaVersion, S
             }
         }
 
-        try
+        // There are 3 cases depending on the version of scalac-scoverage-plugin
+        // * Version 2.0.0 onwards - 3 artifacts, plugin using resolvedScalaVersion
+        // * Version 1.4.2         - 1 artifact, plugin using resolvedScalaVersion falling back to scalaMainVersion
+        // * Version 1.4.1 older   - 1 artifact, plugin using scalaMainVersion
+        //
+        final ArtifactVersion pluginArtifactVersion = new DefaultArtifactVersion(resolvedScalacPluginVersion);
+        if(pluginArtifactVersion.getMajorVersion() >= 2) {
```

Comment:
```suggestion
        final ArtifactVersion pluginArtifactVersion = new DefaultArtifactVersion(resolvedScalacPluginVersion);
        if(pluginArtifactVersion.getMajorVersion() > 2) {
          throw new IllegalArgumentException("Currently Scala version 3 or above are not supported by this Maven plugin.");
        }
        if(pluginArtifactVersion.getMajorVersion() == 2) {
```
I guess something like this would make the failure more evident?

Reply:
> > Given there's no support for Scala 3 at the moment, it would probably be better to implement that explicitly in another PR.
> 
> Correct, but we should still warn right now in the case that someone tries this with Scala 3. Or what is the behavior currently? The worse thing would be silent failure because it will just lead to more issues being created and confusion about why it's not working.

The current behaviour will be to fail to resolve the scoverage compiler plugin for Scala 3 as no plugins are currently published newer than 2.13.10

(Link to MVNRepository [here](https://mvnrepository.com/artifact/org.scoverage/scalac-scoverage-plugin))

This will cause a build failure. It's not as nice as explicitly notifying the user.

Reply:
> I guess something like this would make the failure more evident?

The artifact version in the test you referenced is the version of the scoverage plugin. The scala version is calculated through `resolveScalaVersion()`.

That function will default to an explicitly configured scala version or a search for the scala standard library. If I understand correctly, the Scala 3 standard library uses different coordinates from the earlier versions `org.scala-lang:scala3-library`.

If the user is depending on the scala 3 standard library `resolveScalaVersion()` will fail to detect the scala version resulting in a warning being emitted and the execution of the mojo being skipped.

Reply:
To me that sounds like acceptable behavior - @ckipp01 since an explicit scala-3-warning would need explicit scala-3 detection, I would leave that to a separate fix?

Reply:
> To me that sounds like acceptable behavior - @ckipp01 since an explicit scala-3-warning would need explicit scala-3 detection, I would leave that to a separate fix?

Sure that would be fine, but I'd for sure want to try to get that in before we attempt to do another release.